### PR TITLE
Mutation mask, return haploids, use convert to einops.  

### PR DESCRIPTION
**Pull request by jeffhsu3**
This pull request introduces changes that include updates to mutation handling, the addition of weighting in selection functions, and the introduction of a new `meiosis_only` function. Additionally, the codebase has been cleaned up with minor refactorings and expanded documentation.

### Mutation Handling Enhancements:
* Added a `mutation_index_mask` parameter to `cross`, `double_haploid`, and `_meiosis` functions, allowing fine-grained control over where mutations can occur. Default behavior ensures all loci are mutable if no mask is provided (`chromax/functional.py`). [[1]](diffhunk://#diff-0107b702a68485bc3277e3a659e7269e2ccdf9d6c70089f626d363ba68243698R20) [[2]](diffhunk://#diff-0107b702a68485bc3277e3a659e7269e2ccdf9d6c70089f626d363ba68243698R223-R270) [[3]](diffhunk://#diff-0107b702a68485bc3277e3a659e7269e2ccdf9d6c70089f626d363ba68243698L180-R282)
* Updated mutation logic in `_meiosis` to combine mutation probability and mask constraints (`chromax/functional.py`).

### New Features:
* Introduced a `meiosis_only` function for performing meiosis without crossing, returning haploids for selection. This function mirrors the parameters of `cross` but focuses on meiosis (`chromax/functional.py`).
* Added a `weighting` parameter to the `select` function to enable weighted selection based on traits (`chromax/functional.py`, `chromax/simulator.py`). [[1]](diffhunk://#diff-0107b702a68485bc3277e3a659e7269e2ccdf9d6c70089f626d363ba68243698R292) [[2]](diffhunk://#diff-8fea7bae394eae0cc16311903868d608158833ee7689e9dcffa027b6c17e9ee4R439)

### Simulator Updates:
* Added support for `mutation_mask_index` and `genotype_error` in the `Simulator` class, providing additional flexibility for simulation configurations (`chromax/simulator.py`). [[1]](diffhunk://#diff-8fea7bae394eae0cc16311903868d608158833ee7689e9dcffa027b6c17e9ee4R74-R81) [[2]](diffhunk://#diff-8fea7bae394eae0cc16311903868d608158833ee7689e9dcffa027b6c17e9ee4L243-R254)
* Enhanced `cross` and `double_haploid` methods in `Simulator` to utilize the new mutation mask parameter (`chromax/simulator.py`). [[1]](diffhunk://#diff-8fea7bae394eae0cc16311903868d608158833ee7689e9dcffa027b6c17e9ee4L243-R254) [[2]](diffhunk://#diff-8fea7bae394eae0cc16311903868d608158833ee7689e9dcffa027b6c17e9ee4R341)

### Refactoring and Documentation:
* Replaced manual reshaping with `einops.rearrange` for improved readability and maintainability (`chromax/functional.py`).
* Removed unused assertions and added flexibility to the `conventional_index` function for handling multi-trait GEBV models (`chromax/index_functions.py`).

### Workflow Adjustment:
* Removed Python 3.9 from the CI build matrix in `.github/workflows/build.yml`, focusing on more recent Python versions.

### CODE REVIEW

File: chromax/functional.py [hunk 1]
```diff
@@ -48,39 +54,124 @@ def cross(
         >>> f2.shape
         (50, 1000, 2)
     """
-    parents = parents.reshape(*parents.shape[:3], -1, 2)
+    parents = rearrange(parents, "n p m (pl two) -> n p m pl two", two=2)
+    # parents = parents.reshape(*parents.shape[:3], -1, 2)
```
Comment (younik) [inline chromax/functional.py]: nice! feel free to remove the comment

`p=2` always, right? Can also add, to have a further check:
```diff
- rearrange(parents, "n p m (pl two) -> n p m pl two", two=2)
+ rearrange(parents, "n p m (pl two) -> n p m pl two", two=2, p=2)

File: chromax/functional.py [hunk 2]
```diff
@@ -48,39 +54,124 @@ def cross(
...
     offsprings = _cross(
         parents,
         recombination_vec,
         cross_random_key,
         mutate_random_key,
         mutation_probability,
+        mutation_index_mask,
     )
+    # return rearrange("n m pl -> ")
     return offsprings.reshape(*offsprings.shape[:-2], -1)
```
Comment (younik) [inline chromax/functional.py]: either remove the comment or use rearrange and remove reshape

rearrange should be something like:
```python
rearrange("n m pl two -> n m (pl two)", two=2)

File: chromax/functional.py [hunk 3]
```diff
@@ -48,39 +54,124 @@ def cross(
...
 
 
 @jax.jit
-@partial(jax.vmap, in_axes=(0, None, 0, 0, None))  # parallelize across individuals
-@partial(jax.vmap, in_axes=(0, None, 0, 0, None), out_axes=2)  # parallelize parents
+def meiosis_only(
```
Comment (younik) [inline chromax/functional.py]: might even be called `meiosis` instead of `meiosis_only`

File: chromax/functional.py [hunk 4]
```diff
@@ -48,39 +54,124 @@ def cross(
...
+    mutation_index_mask: Bool[Array, N_MARKERS],
+) -> Haploid:
+    return _meiosis(
+        ind,
+        recombination_vec,
+        cross_random_key,
+        mutate_random_key,
+        mutation_probability,
+        mutation_index_mask,
+    )
```
Comment (younik) [inline chromax/functional.py]: as it is only for vmapping meiosis, for simplicity, you can just do it in the meiosis_only method:

```python
    haploids = jax.vmap(_meiosis, in_axes=(0, None, 0, 0, None, None)(
        ...
   )
```

File: chromax/functional.py [hunk 5]
```diff
@@ -187,6 +289,7 @@ def select(
     population: Population["n"],
     k: int,
     f_index: Callable[[Population["n"]], Float[Array, "n"]],
+    weighting: Float[Array, "n traits"] | None = None,
```
Comment (younik) [inline chromax/functional.py:250]: just a small comment for this:

Originally, I assumed f_index was always outputting a scalar for individual (see typing), so weighting was done inside f_index
But I guess having this is a good utility; can you update the typing of f_index accordingly then?

File: chromax/index_functions.py [hunk 6]
```diff
@@ -67,10 +70,13 @@ def conventional_index(GEBV_model: TraitModel):
...
 
-    def conventional_index_f(pop: Population["n"]) -> Float[Array, "n"]:
+    def conventional_index_f(
+        pop: Population["n"], trait_index: int = 0
+    ) -> Float[Array, "n"]:
         gebv = GEBV_model(pop)
-        assert gebv.shape[-1] == 1
-        return gebv[..., 0]
+        # assert gebv.shape[-1] == 1
+        # return gebv[..., trait_index]
```
Comment (younik) [inline chromax/index_functions.py]: remove comments

File: chromax/simulator.py [hunk 7]
```diff
@@ -69,11 +71,14 @@ def __init__(
...
         recombination_column: str = "RecombRate",
         mutation_probability: float = 0.0,
+        mutation_mask_index: Optional[np.ndarray] = None,
         h2: Optional[np.ndarray] = None,
+        genotype_error: float = 0.0,
         seed: Optional[int] = None,
         device: xc.Device = None,
         backend: Union[str, xc._xla.Client] = None,
     ):
+        # :TODO add genotype error
```
Comment (younik) [inline chromax/simulator.py]: either doing the TODO or removing the genotype error args

File: chromax/simulator.py [hunk 8]
```diff
@@ -290,6 +301,7 @@ def diff_cross_f(
             keys_shape = len(cross_weights), len(population), 2, population.shape[-2]
             keys = jax.random.split(random_key, num=np.prod(keys_shape))
             keys = keys.reshape(*keys_shape, 2)
+            print(population.shape)
```
Comment (younik) [inline chromax/simulator.py]: remove print

**Review by younik — COMMENTED** [review]
Thank you @jeffhsu3, the changes look very useful!

I added some minor comments, before merging :)

File: chromax/index_functions.py (hunk 6 above)
Comment (younik) [reply chromax/index_functions.py]: > remove comments

this review is left

**Review by younik — APPROVED** [review]
Thanks! There is one comment left, and the pre-commit fix, then we can merge!

**Comment by younik** [comment]
That's great, thank you!
